File: 2021/day02/solution.hs
import System.IO (readFile)

type Command = (String, Int)

parseLine :: String -> Command
parseLine line = let [cmd, num] = words line in (cmd, read num)

main :: IO ()
main = do
  content <- readFile "input.txt"
  let commands = map parseLine (lines content)

  print (solvePart1 commands)
  print (solvePart2 commands)

multiplyTuple :: (Int, Int) -> Int
multiplyTuple (h, d) = h * d

solvePart1 :: [Command] -> Int
solvePart1 = multiplyTuple . foldl applyCommand (0, 0)
 where
  applyCommand (h, d) ("forward", x) = (h + x, d)
  applyCommand (h, d) ("down", x) = (h, d + x)
  applyCommand (h, d) ("up", x) = (h, d - x)

solvePart2 :: [Command] -> Int
solvePart2 = multiplyTuple . extractPosition . foldl applyCommandWithAim (0, 0, 0)
 where
  applyCommandWithAim :: (Int, Int, Int) -> Command -> (Int, Int, Int)
  applyCommandWithAim (h, d, aim) ("forward", x) = (h + x, d + aim * x, aim)
  applyCommandWithAim (h, d, aim) ("down", x) = (h, d, aim + x)
  applyCommandWithAim (h, d, aim) ("up", x) = (h, d, aim - x)
  extractPosition (h, d, _) = (h, d)

File: 2021/day02/solution.py
import numpy as np


def parse_line(command, number):
    return (command, int(number))


file_name = "input.txt"
with open(file_name, "r") as file:
    data = [parse_line(*line.split(" ")) for line in file]


def part1(data):
    lookup = {
        "forward": np.array([1, 0]),
        "up": np.array([0, -1]),
        "down": np.array([0, 1]),
    }
    pos = np.array([0, 0])

    for command, number in data:
        # Lookup the number mask and add the scaled mask to pos
        pos += lookup[command] * number

    return pos[0] * pos[1]


def part2(data):
    mask = np.array([1, 0])
    pos = np.array([0, 0])

    for command, number in data:
        if command == "forward":
            # Add to position
            pos += mask * number
        else:
            # Increase or decrease aim
            mask[1] += {"up": -1, "down": 1}[command] * number

    return pos[0] * pos[1]


# Running
print("Part 1:", part1(data))  # 1936494
print("Part 2:", part2(data))  # 1997106066
